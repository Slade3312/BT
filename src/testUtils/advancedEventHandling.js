const KEYBOARD_BACKSPACE_KEYCODE = 8;


/**
 * Inject the following into enzyme node:
 * - keypress, and keyboard events that change input value,
 *   manipulate selection and call keydown, keyup, change
 * - selectionStart, selectionEnd and setSelectionRange
 *
 * Usage:
 * ----
 * const component = mount(...);
 * const input = component.find('input');
 * injectEventHandling(input);
 * input.simulate('keypress', { key: 'a' });
 * input.simulate('keypress', { key: 'Backspace' });
 * input.simulate('keyboard', { keys: ['b', 'c', 'Backspace', 'e' ] });
 *
 * Note:
 * ----
 * This module does not exactly reproduce the default browser behavior,
 * it is simply a close approximation with limited functionality
 *
 * Amongst limitations - async nature of cursor position management
 */
export const injectEventHandling = (testNode) => {
  const originalSimulate = testNode.simulate;

  const domNode = testNode.instance();
  Object.defineProperty(domNode, 'selectionStart', {
    writable: true,
    value: 0,
  });
  Object.defineProperty(domNode, 'selectionEnd', {
    writable: true,
    value: 0,
  });

  Object.defineProperty(domNode, 'setSelectionRange', {
    writable: true,
    value: (selectionStart, selectionEnd) => {
      console.assert(
        selectionStart >= 0 && selectionEnd >= selectionStart && selectionEnd <= domNode.value.length,
        'Attempting to call setSelectionRange with improper interval',
      );

      domNode.selectionStart = selectionStart;
      domNode.selectionEnd = selectionEnd;
    },
  });

  testNode.simulate = (eventType, event, ...args) => {
    if (eventType === 'keyboard') {
      /** shortcut event type for user input simulation */
      const { keys } = event;

      keys.forEach(key => testNode.simulate('keypress', { key }));
    } else if (eventType === 'keypress') {
      /** smart keypress, supporting different keys as well as cursor movement */
      let preventDefault;
      event.preventDefault = () => { preventDefault = true; };
      if (event.key === 'Backspace') event.keyCode = KEYBOARD_BACKSPACE_KEYCODE;

      originalSimulate.call(testNode, 'keydown', event, ...args);
      /** must read instance after keydown, otherwise we miss out on changes done by keydown */
      let { value } = testNode.instance();

      switch (event.key) {
        case 'Backspace':
          /**
           * Handles three possible cases for backspace:
           * 1. if selection is empty and we are at start - we do nothing
           * 2. if selection is empty and we are not at start - we erase one symbol backwards
           * 3. if selection is not empty, we erase selection
           */
          if (domNode.selectionStart === domNode.selectionEnd && domNode.selectionStart > 0) {
            domNode.selectionStart -= 1;
          }

          value = value.slice(0, domNode.selectionStart) + value.slice(domNode.selectionEnd);
          domNode.value = value;
          domNode.setSelectionRange(domNode.selectionStart, domNode.selectionStart);

          if (!preventDefault) {
            testNode.simulate('change', { target: { value } });
          }
          break;
        default:
          /** otherwise we just enter a key */
          value = value.slice(0, domNode.selectionStart) + event.key + value.slice(domNode.selectionEnd);
          domNode.value = value;
          domNode.setSelectionRange(domNode.selectionStart + 1, domNode.selectionStart + 1);


          if (!preventDefault) {
            originalSimulate.call(testNode, 'change', { target: { value } });
          }
          break;
      }
      originalSimulate.call(testNode, 'keypress', event, ...args);
      originalSimulate.call(testNode, 'keyup', event, ...args);

      /**
       * this is default react behavior: if expected value doesn't match actual value,
       * aka onChange modifies field - we reset cursor position to last symbol
       */
      // if (value !== domNode.value) {
      //   console.log('BREAK ALL TO ASHES')
      //   domNode.setSelectionRange(domNode.value.length, domNode.value.length);
      // }
    } else if (eventType === 'paste') {
      let { value } = testNode.instance();

      value = value.slice(0, domNode.selectionStart) + event._value + value.slice(domNode.selectionEnd);
      domNode.value = value;
      const nextCursorPos = domNode.selectionStart + event._value.length;
      domNode.setSelectionRange(nextCursorPos, nextCursorPos);

      originalSimulate.call(testNode, 'change', { target: { value } });
    }

    /** keyboard is our abstract event, no propagation needed */
    if (eventType !== 'keyboard' && eventType !== 'paste') {
      originalSimulate.call(testNode, eventType, event, ...args);
    }
  };
};
