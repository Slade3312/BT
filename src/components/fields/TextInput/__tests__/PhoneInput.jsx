import React from 'react';
import { mount } from 'enzyme';

import { injectEventHandling } from 'testUtils/advancedEventHandling';
import { StatefulPhoneInput } from '../index';


const setupTest = (originalValue = '') => {
  const testFn = jest.fn();
  const component = mount(<StatefulPhoneInput onChange={testFn} />);

  const input = component.find('input');
  injectEventHandling(input);
  const domNode = input.instance();

  input.simulate('keyboard', { keys: originalValue.split('') });

  return { testFn, input, domNode };
};


describe('PhoneInput', () => {
  // console.log = () => {};
  // console.assert = () => {};
  it('Should enter value properly', () => {
    const { domNode } = setupTest('9061234567');
    expect(domNode.value).toEqual('+7 906 123-45-67');
  });

  describe('Inserting digits', () => {
    const shouldInsertDigitAt = (caretPos, endCaretPos = caretPos) => {
      it(`Should insert digit at ${caretPos}${caretPos !== endCaretPos ? ` ${endCaretPos}` : ''}`, () => {
        const { input, domNode } = setupTest('9061234567');

        input.simulate('focus');
        domNode.setSelectionRange(caretPos, endCaretPos);
        input.simulate('keypress', { key: '8' });

        expect([domNode.value, domNode.selectionStart, domNode.selectionEnd]).toMatchSnapshot();
      });
    };

    /**
     * with `+7 906 123-45-67`, test inserting `8` at position defined by index
     */
    shouldInsertDigitAt(0);
    shouldInsertDigitAt(2);
    shouldInsertDigitAt(4);
    shouldInsertDigitAt(5);
    shouldInsertDigitAt(6);
    shouldInsertDigitAt(15);
    shouldInsertDigitAt(16);
  });

  describe('Backspace', () => {
    const shouldEraseDigitsAt = (caretPos, endCaretPos = caretPos) => {
      it(`Should erase digit at ${caretPos}${caretPos !== endCaretPos ? ` ${endCaretPos}` : ''}`, () => {
        const { input, domNode } = setupTest('9061234567');

        input.simulate('focus');
        domNode.setSelectionRange(caretPos, endCaretPos);
        input.simulate('keypress', { key: 'Backspace' });

        expect([domNode.value, domNode.selectionStart, domNode.selectionEnd]).toMatchSnapshot();
      });
    };

    /**
     * with `+7 (906) 123-45-67`, test erasing one symbol with backspace at position defined by index
     */
    shouldEraseDigitsAt(0);
    shouldEraseDigitsAt(1);
    shouldEraseDigitsAt(4);
    shouldEraseDigitsAt(7);

    /* this case fails due to nature of testing environment */
    // shouldEraseDigitsAt(9);
    shouldEraseDigitsAt(8);
    shouldEraseDigitsAt(11);
    shouldEraseDigitsAt(15);
    shouldEraseDigitsAt(16);
  });

  describe('Clipboard', () => {
    const shouldPasteStringAt = (originalValue, pastedValue, caretPos, endCaretPos = caretPos, skip = false) => {
      /** FIXME: fix the following tests, and remove skip feature */
      (skip ? it.skip : it)(`Should paste "${pastedValue}" onto "${originalValue}" string ` +
        ` at ${caretPos}${caretPos !== endCaretPos ? ` ${endCaretPos}` : ''}`, () => {
        const { input, domNode } = setupTest(originalValue);

        input.simulate('focus');
        domNode.setSelectionRange(caretPos, endCaretPos);
        input.simulate('paste', { _value: pastedValue });

        expect([domNode.value, domNode.selectionStart, domNode.selectionEnd]).toMatchSnapshot();
      });
    };


    /**
     * with `+7 906 123-45-67`, test erasing one symbol with backspace at position defined by index
     */
    shouldPasteStringAt('', '+7 906 123-45-67', 0, 0, true);
    shouldPasteStringAt('', '+7 906 123-45-67', 1);
    shouldPasteStringAt('', '+7 906 123-45-67', 3);
    shouldPasteStringAt('', '8 906 123-45-67', 0, 0, true);
    shouldPasteStringAt('', '8 906 123-45-67', 3);
    shouldPasteStringAt('', '906 123-45-67', 0);
    shouldPasteStringAt('', '906 123-45-67', 3);

    shouldPasteStringAt('9096005040', '+7 906 123-45-67', 0);
    shouldPasteStringAt('9096005040', '+7 906 123-45-67', 1);
    shouldPasteStringAt('9096005040', '+7 906 123-45-67', 3);

    shouldPasteStringAt('9096005040', '8 906 123-45-67', 0, 0, true);
    shouldPasteStringAt('9096005040', '8 906 123-45-67', 3, 3, true);
    shouldPasteStringAt('9096005040', '906 123-45-67', 0);
    shouldPasteStringAt('9096005040', '906 123-45-67', 3);
  });
});
