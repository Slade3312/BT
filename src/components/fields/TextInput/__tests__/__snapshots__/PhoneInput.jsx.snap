// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhoneInput Backspace Should erase digit at 0 1`] = `
Array [
  "+7 906 123-45-67",
  0,
  0,
]
`;

exports[`PhoneInput Backspace Should erase digit at 1 1`] = `
Array [
  "+7 906 123-45-67",
  0,
  0,
]
`;

exports[`PhoneInput Backspace Should erase digit at 4 1`] = `
Array [
  "+7 061 234-56-7",
  3,
  3,
]
`;

exports[`PhoneInput Backspace Should erase digit at 7 1`] = `
Array [
  "+7 901 234-56-7",
  5,
  5,
]
`;

exports[`PhoneInput Backspace Should erase digit at 8 1`] = `
Array [
  "+7 906 234-56-7",
  6,
  6,
]
`;

exports[`PhoneInput Backspace Should erase digit at 11 1`] = `
Array [
  "+7 906 124-56-7",
  9,
  9,
]
`;

exports[`PhoneInput Backspace Should erase digit at 15 1`] = `
Array [
  "+7 906 123-45-7",
  13,
  13,
]
`;

exports[`PhoneInput Backspace Should erase digit at 16 1`] = `
Array [
  "+7 906 123-45-6",
  15,
  15,
]
`;

exports[`PhoneInput Clipboard Should paste "+7 906 123-45-67" onto "" string  at 0 1`] = `
Array [
  "+7 906 123-45-67",
  16,
  16,
]
`;

exports[`PhoneInput Clipboard Should paste "+7 906 123-45-67" onto "" string  at 1 1`] = `
Array [
  "+7 906 123-45-67",
  16,
  16,
]
`;

exports[`PhoneInput Clipboard Should paste "+7 906 123-45-67" onto "" string  at 3 1`] = `
Array [
  "+7 906 123-45-67",
  16,
  16,
]
`;

exports[`PhoneInput Clipboard Should paste "+7 906 123-45-67" onto "9096005040" string  at 0 1`] = `
Array [
  "+7 906 123-45-67",
  16,
  16,
]
`;

exports[`PhoneInput Clipboard Should paste "+7 906 123-45-67" onto "9096005040" string  at 1 1`] = `
Array [
  "+7 906 123-45-67",
  16,
  16,
]
`;

exports[`PhoneInput Clipboard Should paste "+7 906 123-45-67" onto "9096005040" string  at 3 1`] = `
Array [
  "+7 906 123-45-67",
  16,
  16,
]
`;

exports[`PhoneInput Clipboard Should paste "8 906 123-45-67" onto "" string  at 0 1`] = `
Array [
  "+7 906 123-45-67",
  16,
  16,
]
`;

exports[`PhoneInput Clipboard Should paste "8 906 123-45-67" onto "" string  at 3 1`] = `
Array [
  "+7 906 123-45-67",
  16,
  16,
]
`;

exports[`PhoneInput Clipboard Should paste "8 906 123-45-67" onto "9096005040" string  at 0 1`] = `
Array [
  "+7 906 123-45-67",
  16,
  16,
]
`;

exports[`PhoneInput Clipboard Should paste "8 906 123-45-67" onto "9096005040" string  at 3 1`] = `
Array [
  "+7 906 123-45-67",
  16,
  16,
]
`;

exports[`PhoneInput Clipboard Should paste "906 123-45-67" onto "" string  at 0 1`] = `
Array [
  "+7 906 123-45-67",
  16,
  16,
]
`;

exports[`PhoneInput Clipboard Should paste "906 123-45-67" onto "" string  at 3 1`] = `
Array [
  "+7 906 123-45-67",
  16,
  16,
]
`;

exports[`PhoneInput Clipboard Should paste "906 123-45-67" onto "9096005040" string  at 0 1`] = `
Array [
  "+7 906 123-45-67",
  16,
  16,
]
`;

exports[`PhoneInput Clipboard Should paste "906 123-45-67" onto "9096005040" string  at 3 1`] = `
Array [
  "+7 906 123-45-67",
  16,
  16,
]
`;

exports[`PhoneInput Inserting digits Should insert digit at 0 1`] = `
Array [
  "+7 890 612-34-56",
  4,
  4,
]
`;

exports[`PhoneInput Inserting digits Should insert digit at 2 1`] = `
Array [
  "+7 890 612-34-56",
  4,
  4,
]
`;

exports[`PhoneInput Inserting digits Should insert digit at 4 1`] = `
Array [
  "+7 980 612-34-56",
  5,
  5,
]
`;

exports[`PhoneInput Inserting digits Should insert digit at 5 1`] = `
Array [
  "+7 908 612-34-56",
  6,
  6,
]
`;

exports[`PhoneInput Inserting digits Should insert digit at 6 1`] = `
Array [
  "+7 906 812-34-56",
  8,
  8,
]
`;

exports[`PhoneInput Inserting digits Should insert digit at 15 1`] = `
Array [
  "+7 906 123-45-68",
  16,
  16,
]
`;

exports[`PhoneInput Inserting digits Should insert digit at 16 1`] = `
Array [
  "+7 906 123-45-67",
  16,
  16,
]
`;
